ikarus/finiteelements/fehelper.hh

```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
         localX[i][j] = x[localView.index(localView.tree().child(j).localIndex(i))[0]];
   return localX;
 }
+
+/**
+ * \brief A function which loops over all the nodes of a tree and performs different actions
+ * for a power node and a leaf node depending on the corresponding functor passed.
+ *
+ * \details This function is inspired from the function Dune::TypeTree::Detail::forEachNode
+ * available in dune/typetree/traversal.hh
+ *
+ * \tparam T Type of tree.
+ * \tparam ChildIndex Type of the index of the child.
+ * \tparam PowerFunc Type of the functor called for a power node.
+ * \tparam LeafFunc Type of the functor called for a leaf node.
+ * \param tree Tree whose nodes are to be looped over.
+ * \param childIndex The index of the child.
+ * \param powerFunc A functor to be called for a power node.
+ * \param leafFunc A functor to be called for a leaf node.
+ */
+template <class T, class ChildIndex, class PowerFunc, class LeafFunc>
+void forEachLeafOrPowerNode(T&& tree, ChildIndex&& childIndex, PowerFunc&& powerFunc, LeafFunc&& leafFunc) {
+  using Tree = std::decay_t<T>;
+  if constexpr (Tree::isLeaf) {
+    leafFunc(tree);
+  } else if constexpr (Tree::isPower) {
+    if constexpr (Tree::template Child<Dune::Indices::_0>::Type::isComposite) {
+      auto indices = std::make_index_sequence<Tree::degree()>{};
+      Dune::Hybrid::forEach(indices, [&](auto i) { forEachLeafOrPowerNode(tree.child(i), i, powerFunc, leafFunc); });
+    } else
+      powerFunc(tree, childIndex);
+  } else {
+    auto indices = std::make_index_sequence<Tree::degree()>{};
+    Dune::Hybrid::forEach(indices, [&](auto i) { forEachLeafOrPowerNode(tree.child(i), i, powerFunc, leafFunc); });
+  }
+}
+
+/**
+ * \brief Get the global flat indices for the provided basis.
+ *
+ * \details The global indices are collected in a FlatInterLeaved order. I.e. x_0, y_0, z_0, ..., x_n, y_n, z_n
+ * This function can handle a scalar basis, power basis, and a composite basis.
+ *
+ * \tparam LocalView Type of the local view
+ *
+ * \param localView Local view of the element.
+ * \param globalIndices Output vector to store global indices.
+ */
+template <typename LocalView>
+void globalFlatIndices(const LocalView& localView, std::vector<typename LocalView::MultiIndex>& globalIndices) {
+  globalIndices.clear();
+  using namespace Dune::Indices;
+
+  /** \brief Functor to obtain global indices of a leaf node. */
+  auto leafOpFunc = [&](auto&& node) {
+    const auto& fe = node.finiteElement();
+    for (size_t i = 0; i < fe.size(); ++i) {
+      globalIndices.push_back(localView.index(node.localIndex(i)));
+    }
+  };
+
+  /** \brief Functor to obtain global indices of a power node. */
+  auto powerOpFunc = [&](auto&& node, auto&& ci) {
+    if constexpr (requires { node.child(ci).finiteElement(); }) {
```

Comment:
Can you elaborate again what both of these branches taking care of? The else branch is the usual `power(lagrange,d)` but for which case is the first used?

Reply:
The `if` branch is used if we have `power(lagrange, d)` for example. The `else` branch is used when we have, for example, `composite(lagrange, power(lagrange))`. Then the finite element can be accessed via `child(Dune::Indices::_1, 0).finiteElement()`.

Reply:
To have this differentiation, I have used `childIndex` in the function `forEachLeafOrPowerNode`. This made having `treePath` redundant there and I have removed it. I have to check if it is possible to access the child correctly with the tree path.

Reply:
Ok thanks for the explaination. But the powerOpFunc should receive only the powerNode and not the rootNode. This complicates everything. You should directly pass the child into this function. Then this only uses the `if` branch and the other boilerplate can be removed, which is unnecessary code duplciation.